check_sympy_version.py

```diff
@@ -0,0 +1,128 @@
+VERSION_PATTERN = r"""
+    v?
+    (?:
+        (?:(?P<epoch>[0-9]+)!)?               # epoch
+        (?P<release>[0-9]+(?:\.[0-9]+)*)      # release segment
+        (?P<post>                             # post release
+            (?:-(?P<post_n1>[0-9]+))
+            |
+            (?:
+                [-_\.]?
+                (?P<post_l>post|rev|r)
+                [-_\.]?
+                (?P<post_n2>[0-9]+)?
+            )
+        )?
+    )
+"""
+
+_regex = re.compile(
+    r"^\s*" + VERSION_PATTERN + r"\s*$",
+    re.VERBOSE | re.IGNORECASE,
+)
+
+
+def get_pyodide_lock():
+    gh_api_url = "https://api.github.com/repos/pyodide/pyodide-recipes/releases/latest"
+    response = requests.get(gh_api_url)
+    response.raise_for_status()
+    release_data = response.json()
+
+    lockfile_asset = next(
+        _ for _ in release_data["assets"] if _["name"] == "pyodide-lock.json"
+    )
+
+    response = requests.get(
+        lockfile_asset["browser_download_url"],
+        headers={"Accept": "application/octet-stream"},
+    )
+    response.raise_for_status()
+
+    lock_data = response.json()
+    return lock_data["packages"]["sympy"]["version"]
+
+
+def get_pypi_version():
+    """Get latest stable SymPy version from PyPI."""
+    response = requests.get("https://pypi.org/pypi/sympy/json")
+    response.raise_for_status()
+    data = response.json()
+
+    # Filter versions to only include stable releases
+    versions = []
+    for version in data["releases"].keys():
+        if _regex.match(version):
+            versions.append(version)
+
+    latest_version = sorted(versions, key=Version)[-1]
+    return latest_version
+
+
+def download_wheel(version):
+    """Download the SymPy wheel for the given version."""
+    response = requests.get(f"https://pypi.org/pypi/sympy/{version}/json")
+    response.raise_for_status()
+    data = response.json()
+
+    wheel_url = None
+    for url_info in data["urls"]:
+        if url_info["filename"].endswith("py3-none-any.whl"):
+            wheel_url = url_info["url"]
+            break
+
+    if not wheel_url:
+        print(f"Error: Could not find appropriate wheel for version {version}")
+        sys.exit(1)
+
+    os.makedirs("custom_wheels", exist_ok=True)
+
+    response = requests.get(wheel_url, stream=True)
+    response.raise_for_status()
+    wheel_path = Path("custom_wheels") / f"sympy-{version}-py3-none-any.whl"
+    with open(wheel_path, "wb") as f:
+        for chunk in response.iter_content(chunk_size=8192):
+            f.write(chunk)
+
+    return wheel_path
+
+
+def main():
+    pyodide_version = get_pyodide_lock()
+    pypi_version = get_pypi_version()
+
+    print(f"SymPy version in Pyodide: {pyodide_version}")
+    print(f"Latest PyPI SymPy version: {pypi_version}")
+
+    should_download = Version(pypi_version) > Version(pyodide_version)
```

Comment:
I don't quite understand why all of this is needed. Would it not make sense to just run `pip download sympy` and always use the latest final release?

I'm not sure why the pyodide version would be relevant at all or is there some advantage in using it from pyodide rather than from PyPI?

Reply:
That makes sense to me, I will refactor the script. TIL that a `pip download` command exists!

The idea is to add the `%pip install sympy` command only when it is needed, as it is redundant otherwise.

Reply:
> The idea is to add the `%pip install sympy` command only when it is needed, as it is redundant otherwise.

I don't understand when/where that command is being run. Is it running in the end users browser? Or is it just run locally as part of building the site?

Reply:
The first one – the command runs in the end user's browser, when the REPL is instantiated.